File: AtCoder/AGC024/D.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

vector<int> getDist(const vector<vector<int>>& g, int start){
    const int n = g.size();
    vector<int> res(n, -1);
    queue<int> qu; qu.push(start);
    res[start] = 0;
    while(!qu.empty()){
        int p = qu.front(); qu.pop();
        for(auto& t : g[p]){
            if(res[t] != -1) continue;
            res[t] = res[p] + 1;
            qu.push(t);
        }
    }
    return res;
}

long long check(const vector<vector<int>>& g, const vector<int>& start){
    auto cur = start;
    vector<int> check(g.size(), 0);
    for(auto& t : cur) check[t] = 1;
    long long res = 0;
    while(!cur.empty()){
        int maxDeg = 1;
        for(auto& t : cur) maxDeg = max<int>(maxDeg, g[t].size()-1);
        res *= maxDeg;
        vector<int> next;
        for(auto& t : cur){
            res += maxDeg - (g[t].size() - 1);
            for(auto& p : g[t]){
                if(check[p]) continue;
                check[p] = 1;
                next.push_back(p);
            }
        }
        cur = next;
    }
    return res;
}

int main(){
    int N;
    while(cin >> N){
        vector<vector<int>> g(N);
        for(int i=0;i<N-1;i++){
            int a, b; cin >> a >> b;
            g[a-1].push_back(b-1);
            g[b-1].push_back(a-1);
        }
        auto d = getDist(g, 0);
        int s = distance(d.begin(), max_element(d.begin(), d.end()));
        d = getDist(g, s);
        s = distance(d.begin(), max_element(d.begin(), d.end()));
        vector<int> cur;
        const int maxD = d[s];
        const int color = (maxD+2)/2;
        while(true){
            if(min(d[s], maxD - d[s]) == color-1){
                cur.push_back(s);
            }
            if(d[s] == 0) break;
            for(auto& t : g[s]){
                if(d[t] + 1 == d[s]){
                    s = t;
                    break;
                }
            }
        }
        long long res = check(g, cur);
        if(cur.size() == 1){
            for(auto& t : g[cur[0]]){
                vector<int> start = {cur[0], t};
                res = min(res, check(g, start));
            }
        }
        cout << color << " " << res << endl;
    }
}

File: AtCoder/AGC024/D.rs
use std::io::Read;

fn solve(g: &Vec<Vec<usize>>, s1: usize, s2: usize) -> (i64, i64) {
    let mut color = vec![-1; g.len()];
    let mut res = 1;
    color[s1] = 1;
    if s2 != !0 { color[s2] = 1; res += 1; }
    let mut cur = 1;
    loop {
        let mut mx = 0;
        for i in 0..g.len() {
            if color[i] != cur { continue }
            let mut cnt = 0;
            for &t in &g[i] {
                if color[t] != -1 { continue; }
                cnt += 1;
                color[t] = cur + 1;
            }
            mx = std::cmp::max(mx, cnt);
        }
        if mx == 0 { break }
        res *= mx;
        if res >= 1 << 60 { return (100000, 1 << 60) }
        cur += 1;
    }
    (cur, res)
}

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).ok();
    let mut it = buf.split_whitespace();
    let n = it.next().unwrap().parse::<usize>().unwrap();
    let mut g = vec![Vec::new(); n];
    for _ in 0..n-1 {
        let a = it.next().unwrap().parse::<usize>().unwrap() - 1;
        let b = it.next().unwrap().parse::<usize>().unwrap() - 1;
        g[a].push(b);
        g[b].push(a);
    }
    let mut res = (100000, 0);
    for i in 0..n {
        res = std::cmp::min(res, solve(&g, i, !0));
        for &t in &g[i] {
            if t < i {
                res = std::cmp::min(res, solve(&g, i, t));
            }
        }
    }
    println!("{} {}", res.0, res.1);
}
